datastore/protocol.go

```diff
@@ -1,163 +1,213 @@
+
+		-- Decoded raw_flags
+		-- Only supports stock flags, custom flags must be handled separately
+		allow_multiple_ratings boolean NOT NULL DEFAULT FALSE,
+		round_negatives boolean NOT NULL DEFAULT TRUE,
+		disable_self_rating boolean NOT NULL DEFAULT FALSE,
+
+		-- Decoded raw_internal_flags
+		-- Only supports stock flags, custom flags must be handled separately
+		use_minimum boolean NOT NULL DEFAULT FALSE,
+		use_maximum boolean NOT NULL DEFAULT FALSE,
+
+		PRIMARY KEY (data_id, slot)
+	);`)
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return
 	}
 
-	// * Allow only users whose PIDs are defined in permission.RecipientIDs
-	if permission.Permission == 2 {
-		if !permission.RecipientIDs.Contains(accessorPID) {
-			return nex.NewError(nex.ResultCodes.DataStore.PermissionDenied, "change_error")
-		}
+	_, err = manager.Database.Exec(`CREATE TABLE datastore.ratings (
+		id serial PRIMARY KEY,
+		data_id numeric(20),
+		slot numeric(3),
+		pid numeric(20),
+		value numeric(20),
+		created_at timestamp DEFAULT (CURRENT_TIMESTAMP AT TIME ZONE 'UTC'),
+		updated_at timestamp DEFAULT (CURRENT_TIMESTAMP AT TIME ZONE 'UTC'),
+		FOREIGN KEY (data_id, slot) REFERENCES datastore.rating_settings(data_id, slot)
+	)`)
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return
 	}
 
-	// * Allow only the owner
-	if permission.Permission == 3 {
-		if !ownerPID.Equals(accessorPID) {
-			return nex.NewError(nex.ResultCodes.DataStore.PermissionDenied, "change_error")
-		}
+	_, err = manager.Database.Exec(`CREATE TABLE datastore.rating_locks (
+		pid numeric(20),
+		data_id numeric(20),
+		slot numeric(3),
+		locked_until timestamp,
+		PRIMARY KEY (pid, data_id, slot),
+		FOREIGN KEY (data_id, slot) REFERENCES datastore.rating_settings(data_id, slot)
+	)`)
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return
 	}
-
-	return nil
-}
-
-// SetDataKeyBase sets the base for the key to be used when uploading standard DataStore objects
-func (c *CommonProtocol) SetDataKeyBase(base string) {
-	// * Just in case someone passes a badly formatted key
-	base = strings.TrimPrefix(base, "/")
-	base = strings.TrimSuffix(base, "/")
-	c.s3DataKeyBase = base
-}
-
-// SetNotifyKeyBase sets the base for the key to be used when uploading DataStore notification data
-func (c *CommonProtocol) SetNotifyKeyBase(base string) {
-	// * Just in case someone passes a badly formatted key
-	base = strings.TrimPrefix(base, "/")
-	base = strings.TrimSuffix(base, "/")
-	c.s3NotifyKeyBase = base
-}
-
-// SetMinIOClient sets the MinIO S3 client
-func (c *CommonProtocol) SetMinIOClient(client *minio.Client) {
-	c.minIOClient = client
-	c.S3Presigner = NewS3Presigner(c.minIOClient)
 }
 
 // NewCommonProtocol returns a new CommonProtocol
 func NewCommonProtocol(protocol datastore.Interface) *CommonProtocol {
+	endpoint := protocol.Endpoint().(*nex.PRUDPEndPoint)
```

Comment:
```suggestion
	endpoint := protocol.Endpoint()
```

This will be needed for HPP support. Don't know if this conversion is being applied elsewhere, but it should be removed accordingly